Introduce k8s compatibility mode to C# codegen

mikhailshilkov opened the pull request:
The flag allows us to (mostly) reuse the C# codegen for `pulumi-kubernetes` without making significant breaking changes in the generated code.
At the same time, these changes zero changes in TF-based providers (tested with AWS and Azure).

pgavlin commented:
Two largish pieces of feedback:

1) Are we confident we don't want to take some of these breaking changes? One of the advantages for end users of moving to the schema-based code generator is greater consistency across packages; this approach does not carry that benefit.
2) I don't like the idea of hard-coding knowledge about particular packages into the code generator: besides being brittle, too much of this will over time result in less readable, less maintainable code. Is there an approach we could take that's more generalized? E.g. adding type-level namespace overrides for C#? I think it's okay to take the pragmatic approach given here in the short term, but I'd prefer not to maintain this longer-term.

mikhailshilkov commented:
@pgavlin Great comments! My opinion below, but I'm looking forward to others too.

> Are we confident we don't want to take some of these breaking changes? 

I already take some breaking changes that I'm sure won't impact the users (e.g. deleting some weird never-used classes, or fixing several bugs with input type usage). Maybe, we may squeeze some of the extra changes, but most of them are quite impactful and e.g. namespace structure change is really massive. I don't see how we can live without the flag or some kind of schema-meta-language unless we are willing to bump the NuGet version to 3.0.

> I don't like the idea of hard-coding knowledge about particular packages into the code generator: besides being brittle, too much of this will over time result in less readable, less maintainable code

Agreed!

>  Is there an approach we could take that's more generalized?

I don't see a single theme behind all these changes except that's how the old version worked. We may come up with several flags instead of one, which may be more meaningful, but I'm not sure I see real benefits.

> I think it's okay to take the pragmatic approach given here in the short term

The term here is a 3.0 release when we can remove the majority of these checks if we are okay to make users upgrade their code.

FWIW, I expect more flags to come once we land TS support for both TF and K8s. Do you see another way around?

lukehoban commented:
Do we have a concrete idea of the impact of a breaking change if we were to make one here?  How disruptive would it be?  It's not clear that the user-facing view that that breaking change would be net better - and would be disruptive.

mikhailshilkov commented:
@lukehoban  I can catalog all the changes if that helps. The biggest ones are:

- All input and output types are moved to different namespaces
- Many input/output properties change nullability
- Some constructor parameters don't have default values anymore

lukehoban commented:
> I can catalog all the changes if that helps.

That would be helpful so we can make an informed decision about whether a significant breaking change is warranted.

I'd love to see the impact on at least one or two "meaningful" examples to understand how impactful these changes are in "common" cases as well.

mikhailshilkov commented:
@lukehoban Which alternatives are we considering?  Do you want to make decisions for each type of change individually, or a binary one "keep all/break all/bump major version"? 

The most obvious change would be that none of these namespaces would exist anymore: https://github.com/pulumi/examples/blob/master/kubernetes-cs-guestbook/simple/Guestbook.cs#L4-L6

Others would probably be "just" nullability warnings in most common cases.

mikhailshilkov commented:
The list of breaking changes if we use standard generation for the k8s SDK:

- All input types (args) are moved from `Pulumi.Kubernetes.Types.Inputs.App.Version` to `Pulumi.Kubernetes.App.Version` or `Pulumi.Kubernetes.App.Version.Inputs`. Users have to adjust their `using` statements and/or explicit namespace usage
- Args classes are de-duped: e.g., currently, one can assign the args of `Pod` to `Items` in `PodList` because both use the same `PodArgs` type. That's not the case for the generic generator: there would be two `PodArgs` classes in two namespaces
- All output types are moved from `Pulumi.Kubernetes.Types.Outputs.App.Version` to `Pulumi.Kubernetes.App.Version.Outputs`
- JSON properties change the type to `Input<object>`, e.g. `ControllerRevisionArgs.Data`
- All classes are marked as sealed, which breaks any derived classes (probably, not a problem in practice)
- All output properties become nullable, e.g. `ApiVersion` changes from `Output<string>` to `Output<string?>`
- `KubernetesResource` base class is gone. Honestly, I would be happy to kill it - the YAML implementation has changed since it was introduced, but we haven't cleaned it up. But that leads to breaking changes in most YAML-related public API.
- Most argument constructors stop having a default value of `null`. Probably, a good change, but in theory - breaking.
- All resources lose the internal ad-hoc dictionary-based constructor. We have to keep it, but we can refactor this to a separate flag with a more generic name.

All these changes are breaking both on source-code and binary (except nullability) levels.

lukehoban commented:
We had an offline discussion yesterday on this.  

While we would directionally like to get to fully standardized code generation, we would only want to take that big breaking change once - which means we need to first do the work to transition all languages and for both tf and kubernetes code generation over so we can fully quantify the breaking change to get to fully unified, and we can make said breaking change all at once.

However, we do not want to hold up progress on that.  The changes in this PR are targeted enough that we should make them now, enable ourselves to move to the new generators without a breaking change, and then independently explore the breaking changes it would take to remove all of these special cases as part of a 3.0 release (or similar) across all of our languages/clouds.

Concretely, we should take this change, and open up another issue to track unifying code generation and removing special cases for Kubernetes (and any other special-cases we add as we migrate all code-generation over).

pgavlin commented:
A few specific points that I think are appropriate to address in a non-k8s-specific fashion:

- > JSON properties change the type to `Input<object>`, e.g. ControllerRevisionArgs.Data

  We should be able to accommodate this in the schema already by using C#-specific (but not k8s-specific) annotations. We might be able to accommodate this for all languages by using named types and overlays.

- > All output properties become nullable, e.g. ApiVersion changes from `Output<string>` to `Output<string?>`

  This seems to indicate that the k8s -> schema translation is going awry: marking these output properties as required sounds like the right thing to do, no? Unless that causes breaking changes in other languages.

- > All resources lose the internal ad-hoc dictionary-based constructor. We have to keep it, but we can refactor this to a separate flag with a more generic name.

  Yeah, it seems like we can do this in a more generic way. Given that these are internal APIs, they're not public breaks in any case. I think we'll want to add a public API for this in the future.


Aside from those changes, I can get behind the idea of a k8s-specific flag as long as we are committed to removing it in the future.

lblackstone commented:
> * This seems to indicate that the k8s -> schema translation is going awry: marking these output properties as required sounds like the right thing to do, no? Unless that causes breaking changes in other languages.

I dug into this the other day and found that the k8s OpenAPI spec actually has these fields marked as optional. Semantically, that doesn't really make sense, but it's faithful to the upstream spec.

lukehoban commented:
> I dug into this the other day and found that the k8s OpenAPI spec actually has these fields marked as optional. Semantically, that doesn't really make sense, but it's faithful to the upstream spec.

That sounds like an unrelated breaking change we are considering making then.  And it seems like a breaking change that is pretty bad for usability.  Are we sure we want to make that breaking change?

mikhailshilkov commented:
To be clear, this PR does not do this breaking change (it has a special case for it). Meanwhile, outputs in Go are already all nullable, so I guess we need to reconcile at some point.

mikhailshilkov commented:
> We should be able to accommodate this in the schema already by using C#-specific (but not k8s-specific) annotations. We might be able to accommodate this for all languages by using named types and overlays.

I believe this is tracked in https://github.com/pulumi/pulumi-terraform-bridge/issues/132. I added a TODO note to remove the current check when that one is done. I'll add a note to the bridge issue above when this PR is merged.

> Yeah, it seems like we can do this in a more generic way. 

I added a separate flag to generate dictionary-based constructors.

mikhailshilkov commented:
> We should be able to accommodate this in the schema already by using C#-specific (but not k8s-specific) annotations. We might be able to accommodate this for all languages by using named types and overlays.

@pgavlin Do you have a specific approach in mind that we used elsewhere?

I don't see anything similar yet. I tried adding a `typeName` property to C#-specific overrides section (`CSharpPropertyInfo`). The problem is that instead of object, we now need to encode JSON - but that's not one C# type: C# needs to gen either `System.Text.Json.JsonElement` for outputs, `Pulumi.InputJson` for inputs, nullable variants of both in case they are optional, array variants of both for collections, etc. This way the schema gets way too C# specific.

Another option it to add something like `"typeHint": "json"` and then leave the input/nullable/array decisions to the schema gen. But shouldn't this then be a provider-level hint (i.e. specified once in schema file). And if yes, how is that beneficial to the kubernetes flag?

Finally, we could just annotate this type as something like `"pulumi.json#/JSON"` instead of `"pulumi.json#/Any"` and deal with it in all languages.

Looking forward to your suggestions!

pgavlin commented:
> Finally, we could just annotate this type as something like "pulumi.json#/JSON" instead of "pulumi.json#/Any" and deal with it in all languages.

This is essentially what I was imagining.

mikhailshilkov commented:
> This is essentially what I was imagining.

You mentioned it to be C#-specific, so I looked into that direction first. I wonder what the potential impact of a new type on other languages would be, and whether "json" makes sense there. I'll take a look.

mikhailshilkov commented:
@pgavlin I added a JSON type which fallbacks to Any in most cases except C# codegen. PTAL.
